spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/metadata/JdbcMetadataStore.java

```diff
@@ -0,0 +1,210 @@
+ */
+
+package org.springframework.integration.jdbc.metadata;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.sql.DataSource;
+
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.integration.metadata.ConcurrentMetadataStore;
+import org.springframework.integration.metadata.MetadataStore;
+import org.springframework.jdbc.core.JdbcOperations;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Jdbc-based implementation of {@link MetadataStore}.
+ *
+ * @author Bojan Vukasovic
+ * @since 5.0
+ */
+public class JdbcMetadataStore implements ConcurrentMetadataStore {
+
+	/**
+	 * Default value for the table prefix property.
+	 */
+	public static final String DEFAULT_TABLE_PREFIX = "INT_";
+
+	private volatile String tablePrefix = DEFAULT_TABLE_PREFIX;
+
+	private final JdbcOperations jdbcTemplate;
+
+	private volatile Map<Query, String> queryCache = new HashMap<>();
+
+	private enum Query {
+		GET_VALUE("SELECT METADATA_VALUE FROM %PREFIX%METADATA_STORE WHERE METADATA_KEY=?"),
+		GET_VALUE_FOR_UPDATE("SELECT METADATA_VALUE FROM %PREFIX%METADATA_STORE WHERE METADATA_KEY=? FOR UPDATE"),
+		REPLACE_VALUE("UPDATE %PREFIX%METADATA_STORE SET METADATA_VALUE=? WHERE METADATA_KEY=? AND METADATA_VALUE=?"),
+		REPLACE_VALUE_BY_KEY("UPDATE %PREFIX%METADATA_STORE SET METADATA_VALUE=? WHERE METADATA_KEY=?"),
+		REMOVE_VALUE("DELETE FROM %PREFIX%METADATA_STORE WHERE METADATA_KEY=?"),
+		PUT_IF_ABSENT_VALUE("INSERT INTO %PREFIX%METADATA_STORE(METADATA_KEY, METADATA_VALUE) "
+				+ "SELECT ?, ? FROM INT_METADATA_STORE WHERE METADATA_KEY=? HAVING COUNT(*)=0");
+
+		private String sql;
+
+		Query(String sql) {
+			this.sql = sql;
+		}
+
+		public String getSql() {
+			return this.sql;
+		}
+	}
+
+	/**
+	 * Instantiate a {@link JdbcMetadataStore} using provided dataSource {@link DataSource}.
+	 * @param dataSource a {@link DataSource}
+	 */
+	public JdbcMetadataStore(DataSource dataSource) {
+		this(new JdbcTemplate(dataSource));
+	}
+
+	/**
+	 * Instantiate a {@link JdbcMetadataStore} using provided jdbcOperations {@link JdbcOperations}.
+	 * @param jdbcOperations a {@link JdbcOperations}
+	 */
+	public JdbcMetadataStore(JdbcOperations jdbcOperations) {
+		Assert.notNull(jdbcOperations, "'jdbcOperations' must not be null");
+		this.jdbcTemplate = jdbcOperations;
+	}
+
+	/**
+	 * Public setter for the table prefix property. This will be prefixed to all the table names before queries are
+	 * executed. Defaults to {@link #DEFAULT_TABLE_PREFIX}.
+	 *
+	 * @param tablePrefix the tablePrefix to set
+	 */
+	public void setTablePrefix(String tablePrefix) {
+		this.tablePrefix = tablePrefix;
+	}
+
+	@Override
+	@Transactional(propagation = Propagation.REQUIRES_NEW)
```

Comment:
In JdbcMetadataStore I used transactions in order to make 2 or more SQL statements atomic (wherever I used FOR UPDATE - in putIfAbsent(), put() and remove()). This way inside transaction I'm sure that row that is selected FOR UPDATE is not modified. I used _NEW just to narrow time that lock is held on db row in this way (FOR UPDATE).

Regarding my other PR, I suppose that if I use same transaction (and DB server) inside MetadataStore, Poller and my class that saves to DB it would work probably. I would't have retry policies and ttl as in my other PR but it would hopefully work.